server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -515,6 +515,8 @@ private <T> void ensureAfterSeqNoRefreshed(
             }
 
             final AtomicBoolean isDone = new AtomicBoolean(false);
+            // TODO: this logic should be improved, the timeout should be handled in a way that removes the listener from the logic in the
+            // index shard on timeout so that a timed-out listener does not use up any listener slots.
```

Comment:
++. I'd like to see this done inside `RefreshListeners`, but we can look at this in a follow-up (as discussed elsewhere).